### PR TITLE
feat: [CDS-41559]: Custom Deployment Template CRUD and listing

### PR DESCRIPTION
**Pull request by samarth-sharma31**
### Summary

Adds support for CRUD operations for Deployment template.

#### Screenshots

NA

https://user-images.githubusercontent.com/96409598/189056755-543bcdd0-f5ef-410c-ab62-b39c06cdb7fb.mov

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/10550)

**Review by yc147 — DISMISSED** [review]

File: src/modules/85-cv/components/Notifications/NotificationsContainer.tsx [hunk 1]
```diff
@@ -8,8 +8,8 @@
 import React from 'react'
 import { Container, Icon, PageError, Text } from '@wings-software/uicore'
 import { Color } from '@harness/design-system'
-import CardWithOuterTitle from '@cv/pages/health-source/common/CardWithOuterTitle/CardWithOuterTitle'
 import { useStrings } from 'framework/strings'
+import CardWithOuterTitle from '@pipeline/components/CardWithOuterTitle/CardWithOuterTitle'
```
Comment (deepesh-ui) [inline src/modules/85-cv/components/Notifications/NotificationsContainer.tsx:12]: if we plan on using  CardWithOuterTitle across modules .can we move it to 10-common  ?

File: src/modules/72-templates-library/pages/TemplatesPage/TemplatesPage.tsx [hunk 2]
```diff
@@ -77,7 +77,7 @@ export default function TemplatesPage(): React.ReactElement {
   const allowedTemplateTypes = getAllowedTemplateTypes(scope, {
     [TemplateType.SecretManager]: !!CUSTOM_SECRET_MANAGER_NG,
     [TemplateType.MonitoredService]: !!CVNG_TEMPLATE_MONITORED_SERVICE,
-    [TemplateType.CustomDeployment]: !!NG_DEPLOYMENT_TEMPLATE
+    [TemplateType.CustomDeployment]: !NG_DEPLOYMENT_TEMPLATE
```
Comment (yc147) [inline src/modules/72-templates-library/pages/TemplatesPage/TemplatesPage.tsx:80]: This has to be `!NG_DEPLOYMENT_TEMPLATE`

File: src/modules/72-templates-library/pages/TemplatesPage/views/NewTemplatePopover/NewTemplatePopover.tsx [hunk 3]
```diff
@@ -38,7 +38,7 @@ function NewTemplatePopoverWrapper(): React.ReactElement {
   const allowedTemplateTypes = getAllowedTemplateTypes(getScopeFromDTO(params), {
     [TemplateType.SecretManager]: !!CUSTOM_SECRET_MANAGER_NG,
     [TemplateType.MonitoredService]: !!CVNG_TEMPLATE_MONITORED_SERVICE,
-    [TemplateType.CustomDeployment]: !!NG_DEPLOYMENT_TEMPLATE
+    [TemplateType.CustomDeployment]: !NG_DEPLOYMENT_TEMPLATE
```
Comment (yc147) [inline src/modules/72-templates-library/pages/TemplatesPage/views/NewTemplatePopover/NewTemplatePopover.tsx:41]: This has to be !!NG_DEPLOYMENT_TEMPLATE

File: src/modules/72-templates-library/pages/TemplatesPage/TemplatesPage.tsx (hunk 2 above)
Comment (PratyushGar) [reply src/modules/72-templates-library/pages/TemplatesPage/TemplatesPage.tsx]: Corrected

**Review by deepesh-ui — APPROVED** [review]

File: src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+    drawerConfig?: {
+      shouldShowApplyChangesBtn: boolean
+    }
+    isDrawerOpen?: boolean
+  }
+}
+
+export interface DeploymentContextProps {
+  onDeploymentConfigUpdate: (configValues: any) => Promise<void>
```
Comment (yc147) [inline src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx:37]: Add proper type

File: src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  isReadOnly: boolean
+  gitDetails: EntityGitDetails
+  queryParams: GetPipelineQueryParams
+  stepsFactory: AbstractStepFactory
+}
+
+export interface DeploymentConfigValues
+  extends Omit<DeploymentContextProps, 'queryParams' | 'deploymentConfigInitialValues' | 'onDeploymentConfigUpdate'> {
+  deploymentConfig: DeploymentConfig
+  updateDeploymentConfig: (configValues: any) => Promise<void>
```
Comment (yc147) [inline src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx:48]: Add proper type

File: src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx [hunk 6]
```diff
@@ -115,13 +118,10 @@ export function TemplateVariablesContextProvider(
...
-      resolvedTemplate.type?.toLowerCase() === monitoredServiceYamlKey.toLowerCase()
-        ? monitoredServiceYamlKey
-        : resolvedTemplate.type?.toLowerCase() === secretManagerYamlKey.toLowerCase()
-        ? secretManagerYamlKey
-        : resolvedTemplate.type?.toLowerCase()
+      templateTypeYamlKeyMap[resolvedTemplate.type?.toLowerCase()] || resolvedTemplate.type?.toLowerCase()
+
     setTemplateVariablesData({
-      metadataMap: defaultTo(data?.data?.metadataMap, {}),
+      metadataMap: defaultTo(data?.data?.metadataMap as YamlExtraProperties, {}),
```
Comment (yc147) [inline src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx:124]: Why are we type casting it to `YamlExtraProperties`?

File: src/modules/70-pipeline/components/DeploymentConfigRuntimeInputs/DeploymentConfigRuntimeInputs.tsx [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+import cx from 'classnames'
+import { useStrings } from 'framework/strings'
+import { CustomVariableInputSet } from '@pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet'
+import type { AllNGVariables } from '@pipeline/utils/types'
+import stepCss from '@pipeline/components/PipelineSteps/Steps/Steps.module.scss'
+import css from './DeploymentConfigRuntimeInputs.module.scss'
+
+interface Props {
+  allowableTypes: AllowedTypes
+  template?: any
```
[the remark this answers is not in the document]
Comment (samarth-sharma31) [reply src/modules/70-pipeline/components/DeploymentConfigRuntimeInputs/DeploymentConfigRuntimeInputs.tsx]: done

File: src/modules/72-templates-library/components/TemplateInputs/TemplateInputs.tsx [hunk 8]
```diff
@@ -197,6 +198,21 @@ export const TemplateInputs: React.FC<TemplateInputsProps> = ({ template }) => {
...
                       )}
+                      {templateEntityType === TemplateType.CustomDeployment && (
+                        <Container
+                          className={css.inputsCard}
+                          background={Color.WHITE}
+                          padding={'large'}
+                          margin={{ bottom: 'xxlarge' }}
+                        >
+                          <DeploymentConfigRuntimeInputs
+                            template={inputSetTemplate as any}
```
[the remark this answers is not in the document]
Comment (samarth-sharma31) [reply src/modules/72-templates-library/components/TemplateInputs/TemplateInputs.tsx]: done

File: src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx (hunk 5 above)
Comment (samarth-sharma31) [reply src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx]: done

File: src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx (hunk 4 above)
Comment (samarth-sharma31) [reply src/modules/75-cd/context/DeploymentContext/DeploymentContextProvider.tsx]: done

File: src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx (hunk 6 above)
Comment (PratyushGar) [reply src/modules/70-pipeline/components/TemplateVariablesContext/TemplateVariablesContext.tsx]: Thanks, have updated the type, was being used before as KVPair to bypass.

**Review by yc147 — APPROVED** [review]

**Comment by samarth-sharma31** [comment]
trigger cypress

**Review by PratyushGar — APPROVED** [review]

**Review by deepesh-ui — APPROVED** [review]

**Review by kapilkumawat86 — APPROVED** [review]

**Pull request merged**